### PR TITLE
feat: add Special Rewards calculation and test

### PR DESCRIPTION
**Pull request by HenryQW**
blocked by https://github.com/RSS3-Network/Global-Indexer/pull/150

Also reading Settler constants from config:
- batch_size
- epoch_interval_in_hours

### CODE REVIEW

File: internal/service/settler/settler.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+	var totalEffectiveStakers uint64
+
+	maxPoolSize := big.NewInt(0)
+
+	for _, node := range nodes {
+		poolSize := new(big.Int)
+
+		if _, ok := poolSize.SetString(node.StakingPoolTokens, 10); !ok {
+			fmt.Errorf("failed to parse staking pool tokens: %s", node.StakingPoolTokens)
```
Comment (kallydev) [inline internal/service/settler/settler.go:182]: Unused error.

File: internal/service/settler/settler.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+		// no stakers, no rewards
+		if stackers == 0 {
+			rewards[i] = big.NewInt(0)
+			continue
+		}
+
+		poolSize := new(big.Int)
+
+		if _, ok := poolSize.SetString(node.StakingPoolTokens, 10); !ok {
+			fmt.Errorf("failed to parse staking pool tokens: %s", node.StakingPoolTokens)
```
Comment (kallydev) [inline internal/service/settler/settler.go:214]: Unused error.

**Review by kallydev — CHANGES_REQUESTED** [review]

File: internal/service/settler/settler.go [hunk 3]
```diff
@@ -127,32 +140,25 @@ func (s *Server) constructSettlementData(ctx context.Context, epoch uint64, curs
...
-	for i := range slice {
-		slice[i] = big.NewInt(0)
+	for _, node := range nodes {
+		if nodeInfo, ok := nodeInfoMap[node.Address]; ok {
+			node.StakingPoolTokens = nodeInfo.StakingPoolTokens.String()
+		}
 	}
 
-	return slice
+	return nil
```
Comment (polebug) [inline internal/service/settler/settler.go:161]: I am confused, why does it return nil?

File: internal/service/settler/server.go [hunk 4]
```diff
@@ -218,6 +219,16 @@ func New(ctx context.Context, databaseClient database.Client, redisClient *redis
 		return nil, fmt.Errorf("get chain ID: %w", err)
 	}
 
+	contractAddresses := l2.ContractMap[chainID.Uint64()]
+	if contractAddresses == nil {
+		return nil, fmt.Errorf("contract address not found for chain id: %d", chainID.Uint64())
+	}
+
+	stakingContract, err := l2.NewStaking(contractAddresses.AddressStakingProxy, ethereumClient)
```
Comment (polebug) [inline internal/service/settler/server.go:227]: `ethereumClient` and `stakingContract` are duplicates

File: internal/service/settler/settler.go [hunk 5]
```diff
@@ -127,32 +140,28 @@ func (s *Server) constructSettlementData(ctx context.Context, epoch uint64, curs
...
 
-	// For Alpha, set the counts to 0
-	for i := range slice {
-		slice[i] = big.NewInt(0)
+	for _, node := range nodes {
+		if nodeInfo, ok := nodeInfoMap[node.Address]; ok {
+			stakePoolTokens := nodeInfo.StakingPoolTokens
+			tokenDecimals := big.NewInt(1e18)
+			stakePoolTokens.Div(stakePoolTokens, tokenDecimals)
+			node.StakingPoolTokens = stakePoolTokens.String()
```
Comment (polebug) [inline internal/service/settler/settler.go:160]: There should be a need to maintain 18 precision, I think

File: internal/service/settler/settler.go [hunk 6]
```diff
@@ -127,32 +140,28 @@ func (s *Server) constructSettlementData(ctx context.Context, epoch uint64, curs
...
+	})
 
-	// For Alpha, set the counts to 0
-	for i := range slice {
-		slice[i] = big.NewInt(0)
+	for _, node := range nodes {
+		if nodeInfo, ok := nodeInfoMap[node.Address]; ok {
+			stakePoolTokens := nodeInfo.StakingPoolTokens
+			tokenDecimals := big.NewInt(1e18)
+			stakePoolTokens.Div(stakePoolTokens, tokenDecimals)
```
Comment (polebug) [inline internal/service/settler/settler.go:159]: It's better to keep 18 decimals here I think, because it is handled this way elsewhere as well

**Review by polebug — DISMISSED** [review]

**Review by pseudoyu — APPROVED** [review]

**Review by polebug — APPROVED** [review]

**Pull request merged**